interfaces/prompting/patterns/render.go

```diff
@@ -0,0 +1,371 @@
+type seq []renderNode
+
+func (n seq) NumVariants() int {
+	num := 1
+
+	for i := range n {
+		num *= n[i].NumVariants()
+	}
+
+	return num
+}
+
+func (n seq) InitialVariant() variantState {
+	v := &seqVariant{
+		seq:      n,
+		elements: make([]variantState, len(n)),
+	}
+
+	for i := range n {
+		v.elements[i] = n[i].InitialVariant()
+	}
+
+	return v
+}
+
+func (n seq) nodeEqual(other renderNode) bool {
+	if other, ok := other.(seq); ok {
+		if len(other) != len(n) {
+			return false
+		}
+
+		for i := range n {
+			if !n[i].nodeEqual(other[i]) {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	return false
+}
+
+// optimize for seq joins adjacent literals into a single node, reduces an
+// empty seq to a literal "", and reduces a seq with one element to that
+// element.
+func (n seq) optimize() (renderNode, error) {
+	var b strings.Builder
+
+	var newSeq seq
+
+	for _, item := range n {
+		if l, ok := item.(literal); ok {
+			if l == "" {
+				continue
+			}
+
+			b.WriteString(string(l))
+		} else {
+			if b.Len() > 0 {
+				newSeq = append(newSeq, literal(b.String()))
+				b.Reset()
+			}
+
+			newSeq = append(newSeq, item)
+		}
+	}
+
+	if b.Len() > 0 {
+		newSeq = append(newSeq, literal(b.String()))
+		b.Reset()
+	}
+
+	// Reduce strength
+	switch len(newSeq) {
+	case 0:
+		return literal(""), nil
+	case 1:
+		return newSeq[0], nil
+	}
+
+	return newSeq, nil
+}
+
+// seqVariant is the variant state for a seqeunce of render nodes.
+//
+// Each render node in the seq has a corresponding variant at the same index.
+// of the elements list.
+type seqVariant struct {
+	seq      seq
+	elements []variantState
+}
+
+func (v *seqVariant) Render(buf *bytes.Buffer, alreadyRendered int) int {
+	for _, element := range v.elements {
+		alreadyRendered = element.Render(buf, alreadyRendered)
+	}
+
+	return alreadyRendered
+}
+
+func (v *seqVariant) NextVariant() (length, lengthUnchanged int, moreRemain bool) {
+	var i int
+	for i = len(v.elements) - 1; i >= 0; i-- {
+		componentLength, componentLengthUnchanged, componentMoreRemain := v.elements[i].NextVariant()
+		if componentMoreRemain {
+			length += componentLength
+			lengthUnchanged = componentLengthUnchanged
+			moreRemain = true
+			break
+		}
+		// Reset the variant state for the node whose variants we just exhausted
+		v.elements[i] = v.seq[i].InitialVariant()
+		// Include the render length of the reset node in the total length
+		length += v.elements[i].Length()
+	}
+	if !moreRemain {
+		// No expansions remain for any node in the sequence
+		return 0, 0, false
+	}
+
+	// We already counted v.elements[i], so count j : 0 <= j < i
+	for j := 0; j < i; j++ {
+		componentLength := v.elements[j].Length()
+		length += componentLength
+		lengthUnchanged += componentLength
+	}
+
+	return length, lengthUnchanged, true
+}
+
+func (v *seqVariant) Length() int {
+	totalLength := 0
+
+	for _, element := range v.elements {
+		totalLength += element.Length()
```

Comment:
looks like this may be getting called mulitple times, any way we could store totalLength in seqVariant to avoid redoing all the the work?

Reply:
I think caching length of the current variant for a `seqVariant` helps, as `literal` and `altVariant` are `O(1)` (`O(depth)` total in variants). More importantly we nearly always if not always need to know the length immediately after calling `InitialVariant`, so we might as well return it as well, meaning we almost never need to call `Length()` directly -- only when we've changed a variant partway through a seq and need to compute the total length again -- but this is cached now, so much faster.